### PR TITLE
Add release build

### PR DESCRIPTION
**Pull request by icywang86rui**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://8teunz1x41.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=361d826e-a86b-4464-a560-08d269124822%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://8teunz1x41.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=025d18a5-7d2b-4d15-a938-678e1d7ea980%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: buildspec-release.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  build:
+    commands:
+      # prepare the release (update versions, changelog etc.)
+      - git-release --prepare
+
+      # run linter
+      - tox -e flake8
+
+      # run unit tests
+      - tox -e py36 test/unit
```
Comment (laurenyu) [inline buildspec-release.yml]: no Python 2?

File: buildspec-release.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      # Create pip archive
+      - build_dir="docker/$FRAMEWORK_VERSION"
+      - python3 setup.py sdist
+      - tar_name=$(ls dist)
+      - cp dist/$tar_name $build_dir
+
+      # Build all images
+      - python3 scripts/build_all.py --account $ACCOUNT --region $AWS_DEFAULT_REGION
+
+      # run local cpu integ tests
```
Comment (laurenyu) [inline buildspec-release.yml]: what about Python 2 and GPU?

File: buildspec.yml [hunk 3]
```diff
@@ -97,15 +97,15 @@ phases:
       # run cpu sagemaker tests
       - |
         if has-matching-changes "test/" "tests/" "src/*.py" "docker/*" "buildspec.yml"; then
-          pytest test/integration/sagemaker --region $AWS_DEFAULT_REGION --docker-base-name $ECR_REPO --account-id $ACCOUNT --tag $CPU_TAG --py-version $CPU_PY_VERSION --instance-type $CPU_INSTANCE_TYPE
+          pytest test/integration/sagemaker -n 8 --region $AWS_DEFAULT_REGION --docker-base-name $ECR_REPO --account-id $ACCOUNT --tag $CPU_TAG --py-version $CPU_PY_VERSION --instance-type $CPU_INSTANCE_TYPE
```
Comment (laurenyu) [inline buildspec.yml]: why 8? (would `auto` make more sense?)

File: scripts/build_all.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+
+args = _parse_args()
+binaries = {
+    'py2-cpu': args.py2_cpu_binary,
+    'py3-cpu': args.py3_cpu_binary,
+    'py2-gpu': args.py2_gpu_binary,
+    'py3-gpu': args.py3_gpu_binary
+}
+build_dir = 'docker/{}'.format(args.version)
```
Comment (laurenyu) [inline scripts/build_all.py]: `os.path`? same for l. 66

File: scripts/build_all.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+# Run docker-login so we can pull the cached image
+login_cmd = subprocess.check_output(
+    'aws ecr get-login --no-include-email --registry-id {}'.format(args.account).split())
+print(login_cmd)
```
Comment (laurenyu) [inline scripts/build_all.py]: I'd preface each command with a description of what is being printed out. same goes for ll. 62 and 69 and similar lines in the other file

File: scripts/build_all.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        print(cmd)
+        subprocess.check_call(cmd.split())
+        tag = '{}-{}-py{}'.format(args.version, arch, py_version)
+        prev_image_uri = '{}.dkr.ecr.{}.amazonaws.com/{}:{}'.format(args.account, args.region, args.repo, tag)
+        build_cmd = 'docker build -f {}/Dockerfile.{} --cache-from {} ' \
+                    '--build-arg py_version={} --build-arg framework_installable={} ' \
+                    '-t {}:{} {}'.format(build_dir, arch, prev_image_uri, py_version, binary_file, args.repo, tag, build_dir)
+        print(build_cmd)
+        subprocess.check_call(build_cmd.split())
+        print('Deleting binary file')
```
Comment (laurenyu) [inline scripts/build_all.py]: I'd add the binary file name to this print statement

File: scripts/publish_all.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+for arch in ['cpu', 'gpu']:
+    for py_version in ['2', '3']:
+        source = '{}:{}-{}-py{}'.format(args.repo, args.version, arch, py_version)
+        dest = '{}.dkr.ecr.{}.amazonaws.com/{}'.format(args.account, args.region, source)
+        tag_cmd = 'docker tag {} {}'.format(source, dest)
+        print(tag_cmd)
+        subprocess.check_call(tag_cmd.split())
+        login_cmd = subprocess.check_output(
+            'aws ecr get-login --no-include-email --registry-id {}'.format(args.account).split())
```
Comment (laurenyu) [inline scripts/publish_all.py]: add the region to this command

File: tox.ini [hunk 8]
```diff
@@ -25,6 +25,7 @@ exclude =
     venv/
     sagemaker-tensorflow-extensions
     benchmarks/
+    scripts/
```
Comment (laurenyu) [inline tox.ini]: why exclude these? there are Python files written by us in that directory

File: buildspec-release.yml (hunk 1 above)
Comment (icywang86rui) [reply buildspec-release.yml]: will add

File: buildspec-release.yml (hunk 2 above)
Comment (icywang86rui) [reply buildspec-release.yml]: Will add

File: buildspec.yml (hunk 3 above)
Comment (icywang86rui) [reply buildspec.yml]: k

File: scripts/build_all.py (hunk 4 above)
Comment (icywang86rui) [reply scripts/build_all.py]: k

File: scripts/build_all.py (hunk 5 above)
Comment (icywang86rui) [reply scripts/build_all.py]: k

File: scripts/build_all.py (hunk 6 above)
Comment (icywang86rui) [reply scripts/build_all.py]: k

File: scripts/publish_all.py (hunk 7 above)
Comment (icywang86rui) [reply scripts/publish_all.py]: will add.

File: tox.ini (hunk 8 above)
Comment (icywang86rui) [reply tox.ini]: will add

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://8teunz1x41.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=3c3eeced-6b5e-431f-bd8f-915f42165bd9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://8teunz1x41.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=6f18cf62-e866-4f75-b0d8-a726cf6e4f32%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request merged**